debezium-storage/debezium-storage-configmap/src/main/java/io/debezium/storage/configmap/ConfigMapFormatter.java

```diff
@@ -0,0 +1,112 @@
+import java.util.Base64;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.ObjectMapper;
+
+import io.debezium.DebeziumException;
+
+/**
+ * This class manage the format of key used for Kubernetes config map
+ * <p>
+ * The keys are converted to a valid config:
+ * alphanumeric characters, '-', '_' or '.' (e.g. 'key.name',  or 'KEY_NAME',  or 'key-name',
+ * regex used for validation is '[-._a-zA-Z0-9]+').
+ * <p>
+ * The values are just base64 encoded
+ *
+ * @author Mario Fiore Vitale
+ */
+public class ConfigMapFormatter {
+
+    public static final String CONNECTOR_NAME_SEPARATOR = ".";
+    public static final String ENTRIES_SEPARATOR = "_";
+    public static final String KEY_VALUE_SEPARATOR = "-";
+    private final ObjectMapper objectMapper;
+
+    public ConfigMapFormatter(ObjectMapper mapper) {
+        this.objectMapper = mapper;
+    }
+
+    public ConfigMapFormatter() {
+        this.objectMapper = new ObjectMapper();
+    }
+
+    public Map<ByteBuffer, ByteBuffer> convertFromStorableFormat(Map<String, String> binaryData) {
+
+        return binaryData.entrySet().stream()
+                .collect(Collectors.toMap(
+                        e -> fromStringKey(e.getKey()),
+                        e -> ByteBuffer.wrap(Base64.getDecoder().decode(e.getValue().getBytes()))));
+    }
+
+    public Map<String, String> convertToStorableFormat(Map<ByteBuffer, ByteBuffer> data) {
+
+        return data.entrySet().stream()
+                .collect(Collectors.toMap(
+                        e -> toStringKey(e.getKey()),
+                        e -> new String(Base64.getEncoder().encode(e.getValue().array()))));
+    }
+
+    private ByteBuffer fromStringKey(String key) {
+
+        String[] split = key.split("\\" + CONNECTOR_NAME_SEPARATOR);
+        String[] entries = split[1].split(ENTRIES_SEPARATOR);
+        Map<String, String> properties = Arrays.stream(entries)
+                .map(this::toEntry)
+                .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
+        try {
+            return ByteBuffer.wrap(objectMapper.writeValueAsBytes(List.of(split[0], properties)));
+        }
+        catch (Exception e) {
+            throw new DebeziumException(String.format("Unable to parse key %s", key), e);
+        }
+    }
+
+    private Map.Entry<String, String> toEntry(String storedEntry) {
+        String[] keyAndValue = storedEntry.split(KEY_VALUE_SEPARATOR);
+        return Map.entry(keyAndValue[0], keyAndValue[1]);
+    }
+
+    private String toStringKey(ByteBuffer key) {
+
+        try {
+            JsonNode jsonNode = objectMapper.readTree(key.array());
+
+            // a valid config key must consist of alphanumeric characters, '-', '_' or '.' (e.g. 'key.name', or 'KEY_NAME', or 'key-name', regex used for validation is '[-._a-zA-Z0-9]+').
```

Comment:
Add a note that offset key is always in a form `[namespace, partition]`. On KC `namespace` is the connector name, for engine it is the configured engine name.

Reply:
Just to be sure for engine/connector name `myengine` and a partitions 

```json
{
  "pkey1": "A1",
  "pkey2": "A2"
}
```

```json
{
  "pkey1": "B1",
  "pkey2": "B2"
}
```

and some arbitrary offsets values (say `{offset: 1}` and `{offset: 2}` the config map would have the following entires?

```text
"engine.pkey1-A1_pkey2-A2": "{offset: 1}"
"engine.pkey1-B1_pkey2-B2": "{offset: 2}" 
```

While this is fine I think we should be able to omit the name of the connector / engine completely here as each connector/engine has it's own config map (which would be actually a reasonable default for the name). This way we can simplify the used separator a bit to something like 

```text
"pkey1-A1.pkey2-A2": "{offset: 1}"
"pkey1-B1.pkey2-B2": "{offset: 2}" 
```

The name of the engine/connector can be placed into annotations.

Reply:
> I think we should be able to omit the name of the connector / engine completely here as each connector/engine has it's own config map (which would be actually a reasonable default for the name). This way we can simplify the used separator a bit to something like

But then you still need to add it when the offset are read from the config map otherwise it will never match the original offset keys. Said that what is the real benefit of this simplification?

Reply:
>  the config map would have the following entires?
>```text
>"engine.pkey1-A1_pkey2-A2": "{offset: 1}"
>"engine.pkey1-B1_pkey2-B2": "{offset: 2}" 
>```

Yes, but more precisely

``text
"myengine.pkey1-A1_pkey2-A2": "e29mZnNldDogMX0="
"myengine.pkey1-B1_pkey2-B2": "e29mZnNldDogMn0=" 
```